graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,327 @@
+package com.apollographql.federation.graphqljava.caching;
+
+import com.apollographql.federation.graphqljava._Entity;
+import graphql.ExecutionResult;
+import graphql.ExecutionResultImpl;
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.InstrumentationState;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.SimpleInstrumentationContext;
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters;
+import graphql.execution.instrumentation.parameters.InstrumentationFieldParameters;
+import graphql.schema.*;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * A GraphQL Java Instrumentation that computes a max age for an operation based
+ * on @cacheControl directives.
+ * <p>
+ * See https://www.apollographql.com/docs/apollo-server/performance/caching/ and
+ * the original implementation at https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-core/src/plugin/cacheControl/index.ts
+ */
+public class CacheControlInstrumentation extends SimpleInstrumentation {
+  private final int defaultMaxAge;
+
+  public static final Object CONTEXT_KEY = new Object();
+
+  private static final String DIRECTIVE_NAME = "cacheControl";
+  private static final String MAX_AGE = "maxAge";
+  private static final String SCOPE = "scope";
+  private static final String INHERIT_MAX_AGE = "inheritMaxAge";
+  private static final String EXTENSION_KEY = "cache-control";
+
+  public CacheControlInstrumentation() {
+    this(0);
+  }
+
+  public CacheControlInstrumentation(int defaultMaxAge) {
+    this.defaultMaxAge = defaultMaxAge;
+  }
+
+  @Override
+  public InstrumentationState createState() {
+    return new CacheControlState();
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginExecution(InstrumentationExecutionParameters parameters) {
+    return new InstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onDispatched(CompletableFuture<ExecutionResult> completableFuture) {}
+
+      @Override
+      public void onCompleted(ExecutionResult executionResult, Throwable throwable) {
+        CacheControlState state = parameters.getInstrumentationState();
+
+        // Attach the policy to the context object
+        state.overallPolicy.maybeAsString().ifPresent(s -> parameters.getGraphQLContext().put(CONTEXT_KEY, s));
```

Comment:
Is it possible for `state.overallPolicy` to actually be null? From what I understand, as soon as there is one field in a query (which should always be the case), there should always be a maxAge? (even if it's the default one)

Reply:
I have tests that assert that the overall policy is null if no field- or type-level policies are present. If we never encounter a directive, nothing happens at all.

Reply:
> if no field- or type-level policies are present.

Shouldn't it use the `defaultMaxAge` in these cases? Because there will always be one root field queried and these have a `defaultMaxAge` ?

Reply:
ah yes, you're correct. it should be "If we never encounter a directive and no defaultMaxAge is set, nothing happens at all."

so this is related to your next question about whether `max-age: 0, public` should be emitted or if it should be null like the TS implementation. @glasser do you have an opinion?

Reply:
Yep, exactly, it's all the same question!